<a href="https://colab.research.google.com/github/hzx365/DatasmithWebsite/blob/main/Pre_Processing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###**0. Preparation**

In [ ]:
import numpy as np
import pandas as pd
from google.colab import drive

In [ ]:
prefix = '/content/drive'
from google.colab import drive
drive.mount(prefix, force_remount=True)

Mounted at /content/drive


In [ ]:
world_cities_path = '/content/drive/My Drive/550_Project_Dataset/worldcities.csv'

udemy_path='/content/drive/My Drive/550_Project_Dataset/Udemy_Course_info.csv'
udemy_review_path='/content/drive/My Drive/550_Project_Dataset/Udemy_Comments.csv'

In [ ]:
glassdoor_path='/content/drive/My Drive/550_Project_Dataset/glassdoor.csv'
glassdoor_review_path='/content/drive/My Drive/550_Project_Dataset/glassdoor_reviews.csv'

In [ ]:
world_cities_data = pd.read_csv(world_cities_path)

In [ ]:
course_info_data = pd.read_csv(udemy_path)
course_comment_data = pd.read_csv(udemy_review_path)

In [ ]:
glassdoor_data = pd.read_csv(glassdoor_path)
glassdoor_review_data = pd.read_csv(glassdoor_review_path)

### **1. World Cities EDA, Pre-processing**




In [ ]:
world_cities_data.shape


(44691, 11)

In [ ]:
world_cities_data.head()

city city_ascii      lat       lng    country iso2 iso3   admin_name  \
0      Tokyo      Tokyo  35.6897  139.6922      Japan   JP  JPN        Tōkyō   
1    Jakarta    Jakarta  -6.1750  106.8275  Indonesia   ID  IDN      Jakarta   
2      Delhi      Delhi  28.6100   77.2300      India   IN  IND        Delhi   
3  Guangzhou  Guangzhou  23.1300  113.2600      China   CN  CHN    Guangdong   
4     Mumbai     Mumbai  19.0761   72.8775      India   IN  IND  Mahārāshtra   

   capital  population          id  
0  primary  37732000.0  1392685764  
1  primary  33756000.0  1360771077  
2    admin  32226000.0  1356872604  
3    admin  26940000.0  1156237133  
4    admin  24973000.0  1356226629

In [ ]:
world_cities_data.columns

Index(['city', 'city_ascii', 'lat', 'lng', 'country', 'iso2', 'iso3',
       'admin_name', 'capital', 'population', 'id'],
      dtype='object')

In [ ]:
world_cities_data.dtypes

city           object
city_ascii     object
lat           float64
lng           float64
country        object
iso2           object
iso3           object
admin_name     object
capital        object
population    float64
id              int64
dtype: object

In [ ]:
for col in world_cities_data.columns:
    if world_cities_data[col].dtype in [np.float64, np.int64] and col not in ['id', 'lat', 'lng']:
        # print(col)
        print(world_cities_data[col].describe())

count    4.438400e+04
mean     1.143739e+05
std      7.148542e+05
min      0.000000e+00
25%      1.218700e+04
50%      2.099900e+04
75%      4.800775e+04
max      3.773200e+07
Name: population, dtype: float64


In [ ]:
for col in world_cities_data.columns:
    if world_cities_data[col].dtype not in [np.float64, np.int64]:
        # print(col)
        print("unique number of values for {}:".format(col), len(world_cities_data[col].unique()))

unique number of values for city: 41343
unique number of values for city_ascii: 41140
unique number of values for country: 241
unique number of values for iso2: 240
unique number of values for iso3: 240
unique number of values for admin_name: 4021
unique number of values for capital: 4


### Data cleaning

####Summary: Capital has ~75% missing data, but we do not need this attribute in our project. We do not need to do any further cleaning on this dataset.

In [ ]:
for col in world_cities_data:

    missing_values_count = world_cities_data[col].isnull().sum()
    missing_value_percentage = missing_values_count/world_cities_data.shape[0]
    # max_value = course_data[col].value_counts().idxmax()
    max_value = world_cities_data[col].mode()

    print(f'{col} \t # of missing values: {missing_values_count} \t % of missing values: {missing_value_percentage}')

city 	 # of missing values: 0 	 % of missing values: 0.0
city_ascii 	 # of missing values: 0 	 % of missing values: 0.0
lat 	 # of missing values: 0 	 % of missing values: 0.0
lng 	 # of missing values: 0 	 % of missing values: 0.0
country 	 # of missing values: 0 	 % of missing values: 0.0
iso2 	 # of missing values: 33 	 % of missing values: 0.0007384037054440491
iso3 	 # of missing values: 0 	 % of missing values: 0.0
admin_name 	 # of missing values: 316 	 % of missing values: 0.007070774876373319
capital 	 # of missing values: 33553 	 % of missing values: 0.7507775614777025
population 	 # of missing values: 307 	 % of missing values: 0.006869392047615851
id 	 # of missing values: 0 	 % of missing values: 0.0


In [ ]:
world_cities_data['capital'].unique()

array(['primary', 'admin', nan, 'minor'], dtype=object)

### Index

In [ ]:
# multi-col index
grouped_cities = world_cities_data.groupby(["lat", "lng", "city"])

counts = grouped_cities.size()
print(counts)

(counts == 1).all()

lat       lng        city             
-54.9333  -67.6167   Puerto Williams      1
-54.8019  -68.3031   Ushuaia              1
-54.2833  -36.5000   King Edward Point    1
-54.2806  -36.5080   Grytviken            1
-53.7833  -67.7000   Río Grande           1
                                         ..
 74.8983   142.1050  Zemlya Bunge         1
 76.0194  -65.1139   Savissivik           1
 77.4667  -69.2306   Qaanaaq              1
 78.2167   15.6333   Longyearbyen         1
 81.7166  -17.8000   Nord                 1
Length: 44691, dtype: int64


True

Below is another attempt to have a single-col index, which wasn't adopted in this project eventually.

In [ ]:
# pick an attribute as the index
# print(len(world_cities_data['id'].unique()))

# print(len(world_cities_data['id']))

# len(world_cities_data['id'].unique()) == len(world_cities_data['id'])


44690
44691


False

In [ ]:
# # find the duplicates
# duplicated_ids = world_cities_data[world_cities_data['id'].duplicated()]

# # Filter the DataFrame to show rows with duplicated IDs
# rows_with_duplicates = world_cities_data[world_cities_data['id'].isin(duplicated_ids['id'])]

# print(rows_with_duplicates)

       city city_ascii      lat      lng country iso2 iso3 admin_name capital  \
7563   Uman       Uman  20.8833 -89.7500  Mexico   MX  MEX    Yucatán   minor   
15401  Umán       Uman  20.8500 -89.7667  Mexico   MX  MEX    Yucatán     NaN   

       population          id  
7563      39611.0  1484009408  
15401     50993.0  1484009408  


In [ ]:
# remove either row
# world_cities_data = world_cities_data.drop(world_cities_data[world_cities_data['city'] == 'Umán'].index)


In [ ]:
# test index again -> index selected
# len(world_cities_data['id'].unique()) == len(world_cities_data['id'])

True

###**Decomposition**



> Useful attributes:

city, lat, lng, country



> 3NF decomposition:

candidate key: lat, lng, city


world_cities (**lat**, **lng**, **city**, country)


FD: lat, lng, city -> country



In [ ]:
### split the table

table1 = world_cities_data[['lat', 'lng', 'city', 'country']].copy()
world_cities_location = table1.to_csv("world_cities.csv", index=False)



###**2. Udemy EDA, Pre-processing**

In [ ]:
course_info_data.shape


(209734, 20)

In [ ]:
course_comment_data.shape

(9411727, 6)

In [ ]:
course_info_data.head()

id                                              title  is_paid   price  \
0  4715             Online Vegan Vegetarian Cooking School     True   24.99   
1  1769         The Lean Startup Talk at Stanford E-Corner    False    0.00   
2  5664  How To Become a Vegan, Vegetarian, or Flexitarian     True   19.99   
3  7723                               How to Train a Puppy     True  199.99   
4  8157                      Web Design from the Ground Up     True  159.99   

                                            headline  num_subscribers  \
0  Learn to cook delicious vegan recipes. Filmed ...             2231   
1  Debunking Myths of Entrepreneurship A startup ...            26474   
2  Get the tools you need for a lifestyle change ...             1713   
3  Train your puppy the right way with Dr. Ian Du...             4988   
4  Learn web design online: Everything you need t...             1266   

   avg_rating  num_reviews  num_comments  num_lectures  content_length_min  \
0        3.75          134            42            37                1268   
1        4.50          709           112             9                  88   
2        4.40           41            13            14                  82   
3        4.80          395            88            36                1511   
4        4.75           38            12            38                 569   

         published_time last_update_date   category          subcategory  \
0  2010-08-05T22:06:13Z          11/6/20  Lifestyle      Food & Beverage   
1  2010-01-12T18:09:46Z              NaN   Business     Entrepreneurship   
2  2010-10-13T18:07:17Z          10/9/19  Lifestyle      Other Lifestyle   
3  2011-06-20T20:08:38Z          1/13/16  Lifestyle  Pet Care & Training   
4  2011-06-23T18:31:20Z              NaN     Design           Web Design   

           topic language                                         course_url  \
0  Vegan Cooking  English           /course/vegan-vegetarian-cooking-school/   
1   Lean Startup  English  /course/the-lean-startup-debunking-myths-of-en...   
2  Vegan Cooking  English  /course/see-my-personal-motivation-for-becomin...   
3   Pet Training  English                /course/complete-dunbar-collection/   
4     Web Design  English             /course/web-design-from-the-ground-up/   

  instructor_name      instructor_url  
0     Angela Poch   /user/angelapoch/  
1       Eric Ries     /user/ericries/  
2     Angela Poch   /user/angelapoch/  
3      Ian Dunbar   /user/ian-dunbar/  
4  E Learning Lab  /user/edwin-ang-2/

In [ ]:
# selected_rows = course_comment_data[course_comment_data['course_id'] == 4715]
# selected_rows

In [ ]:
course_comment_data.head()

id  course_id  rate                       date   display_name  \
0   88962892    3173036   1.0  2021-06-29T18:54:25-07:00          Rahul   
1  125535470    4913148   5.0  2022-10-07T11:17:41-07:00          Marlo   
2   68767147    3178386   3.5  2020-10-19T06:35:37-07:00  Yamila Andrea   
3  125029758    3175814   5.0  2022-09-30T21:13:49-07:00     Jacqueline   
4   76584052    3174896   4.5  2021-01-30T08:45:11-08:00        Anthony   

                                             comment  
0  I think a beginner needs more than you think.\...  
1  Aviva is such a natural teacher and healer/hea...  
2  Muy buena la introducción para entender la bas...  
3  This course is the best on Udemy.  This breakd...  
4  I found this course very helpful. It was full ...

In [ ]:
course_info_data.columns

Index(['id', 'title', 'is_paid', 'price', 'headline', 'num_subscribers',
       'avg_rating', 'num_reviews', 'num_comments', 'num_lectures',
       'content_length_min', 'published_time', 'last_update_date', 'category',
       'subcategory', 'topic', 'language', 'course_url', 'instructor_name',
       'instructor_url'],
      dtype='object')

In [ ]:
course_comment_data.columns

Index(['id', 'course_id', 'rate', 'date', 'display_name', 'comment'], dtype='object')

In [ ]:
course_info_data.dtypes

id                      int64
title                  object
is_paid                  bool
price                 float64
headline               object
num_subscribers         int64
avg_rating            float64
num_reviews             int64
num_comments            int64
num_lectures            int64
content_length_min      int64
published_time         object
last_update_date       object
category               object
subcategory            object
topic                  object
language               object
course_url             object
instructor_name        object
instructor_url         object
dtype: object

In [ ]:
course_comment_data.dtypes

id                int64
course_id         int64
rate            float64
date             object
display_name     object
comment          object
dtype: object

In [ ]:
for col in course_info_data.columns:
    if course_info_data[col].dtype in [np.float64, np.int64] and col != "id":
        # print(col)
        print(course_info_data[col].describe())

count    209734.000000
mean         81.665529
std         117.317846
min           0.000000
25%          19.990000
50%          34.990000
75%          99.990000
max         999.990000
Name: price, dtype: float64
count    2.097340e+05
mean     3.096992e+03
std      1.558132e+04
min      0.000000e+00
25%      2.600000e+01
50%      2.060000e+02
75%      1.435000e+03
max      1.752364e+06
Name: num_subscribers, dtype: float64
count    209734.000000
mean          3.747179
std           1.533711
min           0.000000
25%           3.800000
50%           4.333334
75%           4.625000
max           5.000000
Name: avg_rating, dtype: float64
count    209734.000000
mean        244.358812
std        2458.098276
min           0.000000
25%           3.000000
50%          17.000000
75%          74.000000
max      436457.000000
Name: num_reviews, dtype: float64
count    209734.000000
mean         44.874589
std         355.773107
min           0.000000
25%           1.000000
50%           5.000000
7

In [ ]:
for col in course_comment_data.columns:
    if course_comment_data[col].dtype in [np.float64, np.int64] and col not in ["id", "course_id"]:
        # print(col)
        print(course_comment_data[col].describe())

count    9.411727e+06
mean     4.508870e+00
std      8.587239e-01
min      5.000000e-01
25%      4.000000e+00
50%      5.000000e+00
75%      5.000000e+00
max      5.000000e+00
Name: rate, dtype: float64


In [ ]:
for col in course_info_data.columns:
    if course_info_data[col].dtype not in [np.float64, np.int64]:
        # print(col)
        print("unique number of values for {}:".format(col), len(course_info_data[col].unique()))

unique number of values for title: 208256
unique number of values for is_paid: 2
unique number of values for headline: 202698
unique number of values for published_time: 209562
unique number of values for last_update_date: 3533
unique number of values for category: 13
unique number of values for subcategory: 130
unique number of values for topic: 3819
unique number of values for language: 79
unique number of values for course_url: 209734
unique number of values for instructor_name: 72728
unique number of values for instructor_url: 73515


In [ ]:
for col in course_comment_data.columns:
    if course_comment_data[col].dtype not in [np.float64, np.int64]:
        # print(col)
        print("unique number of values for {}:".format(col), len(course_comment_data[col].unique()))

unique number of values for date: 9138397
unique number of values for display_name: 1420902
unique number of values for comment: 7670271


### Data cleaning

Summary: The two original tables have less than 0.01% of missing values across all attributes.  We do not need to do any further cleaning on this dataset.

In [ ]:
for col in course_info_data:

    missing_values_count = course_info_data[col].isnull().sum()
    missing_value_percentage = missing_values_count/course_info_data.shape[0]
    # max_value = course_data[col].value_counts().idxmax()
    max_value = course_info_data[col].mode()

    print(f'{col} \t # of missing values: {missing_values_count} \t % of missing values: {missing_value_percentage}')

id 	 # of missing values: 0 	 % of missing values: 0.0
title 	 # of missing values: 0 	 % of missing values: 0.0
is_paid 	 # of missing values: 0 	 % of missing values: 0.0
price 	 # of missing values: 0 	 % of missing values: 0.0
headline 	 # of missing values: 27 	 % of missing values: 0.00012873449226162662
num_subscribers 	 # of missing values: 0 	 % of missing values: 0.0
avg_rating 	 # of missing values: 0 	 % of missing values: 0.0
num_reviews 	 # of missing values: 0 	 % of missing values: 0.0
num_comments 	 # of missing values: 0 	 % of missing values: 0.0
num_lectures 	 # of missing values: 0 	 % of missing values: 0.0
content_length_min 	 # of missing values: 0 	 % of missing values: 0.0
published_time 	 # of missing values: 0 	 % of missing values: 0.0
last_update_date 	 # of missing values: 137 	 % of missing values: 0.0006532083496238092
category 	 # of missing values: 0 	 % of missing values: 0.0
subcategory 	 # of missing values: 0 	 % of missing values: 0.0
topic 	 # o

In [ ]:
for col in course_comment_data:

    missing_values_count = course_comment_data[col].isnull().sum()
    missing_value_percentage = missing_values_count/course_comment_data.shape[0]
    # max_value = course_data[col].value_counts().idxmax()
    max_value = course_comment_data[col].mode()

    print(f'{col} \t # of missing values: {missing_values_count} \t % of missing values: {missing_value_percentage}')

id 	 # of missing values: 0 	 % of missing values: 0.0
course_id 	 # of missing values: 0 	 % of missing values: 0.0
rate 	 # of missing values: 0 	 % of missing values: 0.0
date 	 # of missing values: 0 	 % of missing values: 0.0
display_name 	 # of missing values: 75362 	 % of missing values: 0.008007244579023595
comment 	 # of missing values: 6333 	 % of missing values: 0.0006728839457413076


### Index

In [ ]:
# index = 'id' in course_info_data

len(course_info_data['id'].unique()) == len(course_info_data['id'])

True

In [ ]:
# index = 'id' in course_comment_data
len(course_comment_data['id'].unique())==len(course_comment_data['id'])

True

###**Decomposition**

**course_info_data:**


> Useful attributes:


id, title, price, num_subscribers, avg_rating, num_reviews, num_comments, num_lectures, content_length_min, published_time, category, language, instructor_name



> 3NF decomposition:



candidate keys: id

CourseDetail (**id**, title, price, num_subscribers, avg_rating, num_reviews, num_comments, num_lectures, content_length_min, published_time, category, language, instructor_name)

FD: id->title

id->price

id-> num_subscribers

id-> avg_rating

id -> num_reviews,

id-> num_comments

id-> num_lectures

id -> content_length_min

id -> published_time

id-> category

id -> language

id -> instructor_name



**course_comment_data:**

> Useful attributes:


id, course_id, rate, comment


> 3NF decomposition:

candidate key: id

CourseComment (**id**, course_id, rate, comment)

[foreign key course_id references course_info_data.id]

FD: id->course_id

id->rate

id->comment



In [ ]:
### split the table

table1 = course_info_data[['id', 'title', 'price', 'num_subscribers', 'avg_rating', 'num_reviews',
                           'num_comments', 'num_lectures', 'content_length_min', 'published_time',
                           'category', 'language', 'instructor_name']].copy()
course_detail = table1.to_csv("course_info.csv", index=False)


In [ ]:
table1.shape

(209734, 13)

In [ ]:
table4 = course_comment_data[['id', 'course_id', 'rate', 'comment']].copy()
course_comment = table4.to_csv("course_comment.csv", index=False)

In [ ]:
table4.shape

(9411727, 4)

###**3. Glassdoor EDA, Pre-processing**

In [ ]:
glassdoor_data.shape

(165290, 163)

In [ ]:
glassdoor_review_data.shape

(422180, 28)

In [ ]:
glassdoor_data.head()

benefits.benefitRatingDecimal  benefits.comments  benefits.highlights  \
0                            0.0             2801.0               2801.0   
1                            4.0             2802.0               2802.0   
2                            0.0             2803.0               2803.0   
3                            0.0             2804.0               2804.0   
4                            0.0             2805.0               2805.0   

   benefits.numRatings benefits.employerSummary  breadCrumbs  \
0                    0                      NaN         2801   
1                    3                      NaN         2802   
2                    0                      NaN         2803   
3                    0                      NaN         2804   
4                    0                      NaN         2805   

   gaTrackerData.category  gaTrackerData.empId  gaTrackerData.empName  \
0                      -1              2183810            Modus Group   
1                      -1                10416   Boehringer Ingelheim   
2                      -1              1043373        Immobiliare.it    
3                      -1              2226886                  Codup   
4                      -1               851555  Markelytics Solutions   

  gaTrackerData.empSize  ...  salary.currency.displayName salary.currency.id  \
0              501-1000  ...                          NaN                NaN   
1              10000--1  ...   United Kingdom Pound (GBP)                2.0   
2               201-500  ...                          NaN                NaN   
3                Jan-50  ...                          NaN                NaN   
4               201-500  ...                          NaN                NaN   

   salary.currency.name  salary.currency.negativeTemplate  \
0                   NaN                               NaN   
1  United Kingdom Pound                              $(#)   
2                   NaN                               NaN   
3                   NaN                               NaN   
4                   NaN                               NaN   

   salary.currency.new salary.currency.positiveTemplate  \
0                  NaN                              NaN   
1                False                               $#   
2                  NaN                              NaN   
3                  NaN                              NaN   
4                  NaN                              NaN   

  salary.currency.symbol  salary.lastSalaryDate salary.salaries wwfu  
0                    NaN                    NaN          2483.0  NaN  
1                &pound;    2019-10-14T06:05:58          2484.0  NaN  
2                    NaN                    NaN          2485.0  NaN  
3                    NaN                    NaN          2486.0  NaN  
4                    NaN                    NaN          2487.0  NaN  

[5 rows x 163 columns]

In [ ]:
glassdoor_review_data.head()

id  index                                   reviews.val.cons  \
0  44001    0.0               Still not big enough in market place   
1  44002    NaN                                                NaN   
2  44003    NaN                                                NaN   
3  44004    0.0             Nothing important on my point of view.   
4  44004    1.0  Alot of friends working together which isn't v...   

  reviews.val.date reviews.val.featured  reviews.val.helpfulCount  \
0     Oct 20, 2010                False                       0.0   
1              NaN                  NaN                       NaN   
2              NaN                  NaN                       NaN   
3      4 weeks ago                False                       0.0   
4     May 29, 2019                False                       0.0   

   reviews.val.id                                   reviews.val.pros  \
0        689142.0  Great brand , Good leadership , Clear business...   
1             NaN                                                NaN   
2             NaN                                                NaN   
3      29659857.0  Learn new technologies, helpful people, good m...   
4      26381884.0      Very good opportunities to learn technologies   

  reviews.val.publishedOn          reviews.val.publisher  ...  \
0            Oct 20, 2010                       Director  ...   
1                     NaN                            NaN  ...   
2                     NaN                            NaN  ...   
3             4 weeks ago  Software Engineer(Internship)  ...   
4            May 29, 2019               Network Engineer  ...   

    reviews.val.reviewerJobTitle  reviews.val.reviewerLocation  \
0                       Director                 Tokyo (Japan)   
1                            NaN                           NaN   
2                            NaN                           NaN   
3  Software Engineer(Internship)            Alexandria (Egypt)   
4               Network Engineer            Alexandria (Egypt)   

   reviews.val.reviewerStatus  reviews.val.summaryPoints.ceoApproval  \
0             Former Employee                                    1.0   
1                         NaN                                    NaN   
2                         NaN                                    NaN   
3             Former Employee                                    1.0   
4            Current Employee                                    1.0   

   reviews.val.summaryPoints.outlook  reviews.val.summaryPoints.recommend  \
0                                0.0                                  1.0   
1                                NaN                                  NaN   
2                                NaN                                  NaN   
3                                1.0                                  1.0   
4                                1.0                                  1.0   

                                  reviews.val.title  \
0  American Express Japan - Excellent place to work   
1                                               NaN   
2                                               NaN   
3            one of the best places to be intern in   
4                                 Very Good Company   

  reviews.val.adviceToManagement reviews.val.companyResponse  \
0                           None                         NaN   
1                            NaN                         NaN   
2                            NaN                         NaN   
3                            NaN                         NaN   
4                            NaN                         NaN   

  reviews.val.reviewResponses  
0                         NaN  
1                         NaN  
2                         NaN  
3                         NaN  
4                         NaN  

[5 rows x 28 columns]

In [ ]:
glassdoor_data.columns

Index(['benefits.benefitRatingDecimal', 'benefits.comments',
       'benefits.highlights', 'benefits.numRatings',
       'benefits.employerSummary', 'breadCrumbs', 'gaTrackerData.category',
       'gaTrackerData.empId', 'gaTrackerData.empName', 'gaTrackerData.empSize',
       ...
       'salary.currency.displayName', 'salary.currency.id',
       'salary.currency.name', 'salary.currency.negativeTemplate',
       'salary.currency.new', 'salary.currency.positiveTemplate',
       'salary.currency.symbol', 'salary.lastSalaryDate', 'salary.salaries',
       'wwfu'],
      dtype='object', length=163)

In [ ]:
glassdoor_review_data.columns

Index(['id', 'index', 'reviews.val.cons', 'reviews.val.date',
       'reviews.val.featured', 'reviews.val.helpfulCount', 'reviews.val.id',
       'reviews.val.pros', 'reviews.val.publishedOn', 'reviews.val.publisher',
       'reviews.val.reviewRatings.careerOpportunities',
       'reviews.val.reviewRatings.compBenefits',
       'reviews.val.reviewRatings.cultureValues',
       'reviews.val.reviewRatings.overall',
       'reviews.val.reviewRatings.seniorManagement',
       'reviews.val.reviewRatings.worklifeBalance',
       'reviews.val.reviewerDuration', 'reviews.val.reviewerInformation',
       'reviews.val.reviewerJobTitle', 'reviews.val.reviewerLocation',
       'reviews.val.reviewerStatus', 'reviews.val.summaryPoints.ceoApproval',
       'reviews.val.summaryPoints.outlook',
       'reviews.val.summaryPoints.recommend', 'reviews.val.title',
       'reviews.val.adviceToManagement', 'reviews.val.companyResponse',
       'reviews.val.reviewResponses'],
      dtype='object')

In [ ]:
glassdoor_data.dtypes

benefits.benefitRatingDecimal       float64
benefits.comments                   float64
benefits.highlights                 float64
benefits.numRatings                   int64
benefits.employerSummary             object
                                     ...   
salary.currency.positiveTemplate     object
salary.currency.symbol               object
salary.lastSalaryDate                object
salary.salaries                     float64
wwfu                                float64
Length: 163, dtype: object

In [ ]:
print(glassdoor_data['gaTrackerData.pageRequestGuid.guid'].dtype)

object


In [ ]:
print(glassdoor_data['gaTrackerData.jobTitle'].dtype)

object


In [ ]:
print(glassdoor_data['gaTrackerData.jobId.long'].dtype)

float64


In [ ]:
print(glassdoor_data['gaTrackerData.location'].dtype)

object


In [ ]:
print(glassdoor_data['map.country'].dtype)

object


In [ ]:
print(glassdoor_data['job.discoverDate'].dtype)

object


In [ ]:
glassdoor_review_data.dtypes

id                                                 int64
index                                            float64
reviews.val.cons                                  object
reviews.val.date                                  object
reviews.val.featured                              object
reviews.val.helpfulCount                         float64
reviews.val.id                                   float64
reviews.val.pros                                  object
reviews.val.publishedOn                           object
reviews.val.publisher                             object
reviews.val.reviewRatings.careerOpportunities    float64
reviews.val.reviewRatings.compBenefits           float64
reviews.val.reviewRatings.cultureValues          float64
reviews.val.reviewRatings.overall                float64
reviews.val.reviewRatings.seniorManagement       float64
reviews.val.reviewRatings.worklifeBalance        float64
reviews.val.reviewerDuration                      object
reviews.val.reviewerInformation

In [ ]:
glassdoor_useful_cols = ['gaTrackerData.pageRequestGuid.guid','gaTrackerData.jobId.long','gaTrackerData.jobTitle','gaTrackerData.location',
                        'map.country', 'gaTrackerData.sector', 'benefits.benefitRatingDecimal', 'job.description',
                        'header.employerName', 'reviews', 'job.discoverDate']

glassdoor_review_useful_cols = ['id', 'reviews.val.id', 'reviews.val.publishedOn', 'reviews.val.pros', 'reviews.val.cons']

In [ ]:
for col in glassdoor_useful_cols:
    if glassdoor_data[col].dtype in [np.float64, np.int64] and col not in ['gaTrackerData.pageRequestGuid.guid', 'gaTrackerData.jobId.long']:
        # print(col)
        print(glassdoor_data[col].describe())

count    165290.000000
mean          1.387690
std           1.826381
min           0.000000
25%           0.000000
50%           0.000000
75%           3.504717
max           5.000000
Name: benefits.benefitRatingDecimal, dtype: float64
count    165290.000000
mean      82645.500000
std       47715.257335
min           1.000000
25%       41323.250000
50%       82645.500000
75%      123967.750000
max      165290.000000
Name: reviews, dtype: float64


In [ ]:
for col in glassdoor_review_useful_cols:
    if glassdoor_review_data[col].dtype in [np.float64, np.int64] and col not in ['id', 'reviews.val.id']:
        # print(col)
        print(glassdoor_data[col].describe())
# no output, as all cols (except id) are nominal.

### Data cleaning

In [ ]:
for col in glassdoor_useful_cols:

    missing_values_count = glassdoor_data[col].isnull().sum()
    missing_value_percentage = missing_values_count/glassdoor_data.shape[0]
    # max_value = course_data[col].value_counts().idxmax()
    max_value = glassdoor_data[col].mode()

    print(f'{col} \t # of missing values: {missing_values_count} \t % of missing values: {missing_value_percentage}')

gaTrackerData.pageRequestGuid.guid 	 # of missing values: 0 	 % of missing values: 0.0
gaTrackerData.jobId.long 	 # of missing values: 246 	 % of missing values: 0.001488293302680138
gaTrackerData.jobTitle 	 # of missing values: 0 	 % of missing values: 0.0
gaTrackerData.location 	 # of missing values: 70283 	 % of missing values: 0.42521023655393553
map.country 	 # of missing values: 50028 	 % of missing values: 0.302668037993829
gaTrackerData.sector 	 # of missing values: 47373 	 % of missing values: 0.28660536027587874
benefits.benefitRatingDecimal 	 # of missing values: 0 	 % of missing values: 0.0
job.description 	 # of missing values: 4 	 % of missing values: 2.4199891100490048e-05
header.employerName 	 # of missing values: 3245 	 % of missing values: 0.01963216165527255
reviews 	 # of missing values: 0 	 % of missing values: 0.0
job.discoverDate 	 # of missing values: 0 	 % of missing values: 0.0


In [ ]:
# gaTrackerData.location
# glassdoor_missing_rows = glassdoor_data[glassdoor_data['gaTrackerData.location'].isna()]
glassdoor_data_drop_loc = glassdoor_data.dropna(subset=['gaTrackerData.location'])


In [ ]:
glassdoor_data_drop_loc.shape

(95007, 163)

In [ ]:
glassdoor_data_drop_loc_sec = glassdoor_data_drop_loc.dropna(subset= ['gaTrackerData.sector'])

In [ ]:
glassdoor_data_drop_loc_sec.shape

(62681, 163)

In [ ]:
glassdoor_data_drop_loc_sec_country = glassdoor_data_drop_loc_sec.dropna(subset= ['map.country'])

In [ ]:
glassdoor_data_drop_loc_sec_country.shape

(59565, 163)

In [ ]:
glassdoor_data_cleaned = glassdoor_data_drop_loc_sec_country

In [ ]:
glassdoor_data_cleaned.shape

(59565, 163)

#### Summary: glassdoor_data

In the glassdoor_data dataset, the columns of gaTrackerData.location (42% missing value), gaTrackerData.sector (28% missing value), and map.country (30% missing value) contains a high percentage of missing values.

After removing rows that contains missing values in those three columns, the glassdoor_data_cleaned still has 59.565 rows, satisfying the requirements of the 'large dataset' in this project.

The rest of the columns that will be used in this project have less than 0.02% missing value, so we ignore them.




In [ ]:
for col in glassdoor_review_useful_cols:
    if glassdoor_review_data[col].dtype not in [np.float64, np.int64]:
        # print(col)
        print("unique number of values for {}:".format(col), len(glassdoor_review_data[col].unique()))

unique number of values for reviews.val.publishedOn: 3030
unique number of values for reviews.val.pros: 79292
unique number of values for reviews.val.cons: 78651


In [ ]:
for col in glassdoor_review_useful_cols:

    missing_values_count = glassdoor_review_data[col].isnull().sum()
    missing_value_percentage = missing_values_count/glassdoor_review_data.shape[0]
    # max_value = course_data[col].value_counts().idxmax()
    max_value = glassdoor_review_data[col].mode()

    print(f'{col} \t # of missing values: {missing_values_count} \t % of missing values: {missing_value_percentage}')

id 	 # of missing values: 0 	 % of missing values: 0.0
reviews.val.id 	 # of missing values: 78959 	 % of missing values: 0.18702686058079493
reviews.val.publishedOn 	 # of missing values: 78959 	 % of missing values: 0.18702686058079493
reviews.val.pros 	 # of missing values: 78959 	 % of missing values: 0.18702686058079493
reviews.val.cons 	 # of missing values: 78966 	 % of missing values: 0.18704344118622387


In [ ]:
glassdoor_review_data_drops = glassdoor_review_data.dropna(subset=['reviews.val.id','reviews.val.publishedOn','reviews.val.pros', 'reviews.val.cons' ])

In [ ]:
glassdoor_review_data_drops.shape

(343214, 28)

In [ ]:
glassdoor_review_data_drops['reviews.val.publishedOn'] = pd.to_datetime(glassdoor_review_data_drops['reviews.val.publishedOn'], errors='coerce')

<ipython-input-67-0f7510dbf73c>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  glassdoor_review_data_drops['reviews.val.publishedOn'] = pd.to_datetime(glassdoor_review_data_drops['reviews.val.publishedOn'], errors='coerce')


In [ ]:
glassdoor_review_data_drops = glassdoor_review_data_drops.dropna(subset=['reviews.val.publishedOn'])

In [ ]:
glassdoor_review_data_cleaned = glassdoor_review_data_drops

#### Summary: glassdoor_review_data

In the glassdoor_review_data, the columns of reviews.val.publishedOn,reviews.val.pros, reviews.val.cons all have 18% missing data. After removing the missing values, we still have 343,214 rows of data, satisfying the 'big dataset' requirement of this project.  

### Entity resolution

mapping glassdoor_data category to course_info_data category, so that they have the same column to join.

In [ ]:
glassdoor_data_cleaned["gaTrackerData.sector"].unique()

array(['Finance', 'Business Services', 'Information Technology',
       'Accounting & Legal', 'Manufacturing', 'Non-Profit',
       'Oil, Gas, Energy & Utilities', 'Biotech & Pharmaceuticals',
       'Construction, Repair & Maintenance', 'Insurance', 'Education',
       'Retail', 'Telecommunications', 'Media', 'Agriculture & Forestry',
       'Health Care', 'Transportation & Logistics', 'Real Estate',
       'Travel & Tourism', 'Government', 'Aerospace & Defense',
       'Arts, Entertainment & Recreation',
       'Restaurants, Bars & Food Services', 'Consumer Services',
       'Mining & Metals'], dtype=object)

In [ ]:
len(glassdoor_data_cleaned["gaTrackerData.sector"].unique())

25

In [ ]:
course_info_data['category'].unique()

array(['Lifestyle', 'Business', 'Design', 'Health & Fitness',
       'Finance & Accounting', 'Development', 'Marketing',
       'Teaching & Academics', 'IT & Software', 'Office Productivity',
       'Music', 'Personal Development', 'Photography & Video'],
      dtype=object)

In [ ]:
len(course_info_data['category'].unique())

13

In [ ]:
category_mapping = {

    'Finance': 'Finance & Accounting',
    'Business Services': 'Business',
    'Information Technology': 'IT & Software',
    'Accounting & Legal': 'Finance & Accounting',
    'Manufacturing': 'Business',
    'Non-Profit': 'Development',
    'Oil, Gas, Energy & Utilities': 'Business',
    'Biotech & Pharmaceuticals': 'Business',
    'Construction, Repair & Maintenance':'Business',
    'Insurance': 'Finance & Accounting',
    'Education': 'Teaching & Academics',
    'Retail': 'Marketing',
    'Telecommunications': 'IT & Software',
    'Media': 'Photography & Video',
    'Agriculture & Forestry': 'Business',
    'Health Care': 'Health & Fitness',
    'Transportation & Logistics': 'Business',
    'Real Estate': 'Business',
    'Travel & Tourism': 'Lifestyle',
    'Government': 'Business',
    'Aerospace & Defense': 'Business',
    'Arts, Entertainment & Recreation': 'Lifestyle',
    'Restaurants, Bars & Food Services': 'Lifestyle',
    'Consumer Services': 'Marketing',
    'Mining & Metals': 'Business'

}

glassdoor_data_cleaned["gaTrackerData.sector"] = glassdoor_data_cleaned["gaTrackerData.sector"].map(category_mapping).fillna(glassdoor_data_cleaned["gaTrackerData.sector"])

# glassdoor_data.to_csv('updated_glassdoor.csv', index=False)
# glassdoor_data["gaTrackerData.sector"].to_csv('new.csv')
glassdoor_data_cleaned["gaTrackerData.sector"]

<ipython-input-74-7ef39b922448>:31: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  glassdoor_data_cleaned["gaTrackerData.sector"] = glassdoor_data_cleaned["gaTrackerData.sector"].map(category_mapping).fillna(glassdoor_data_cleaned["gaTrackerData.sector"])


0         Finance & Accounting
4                     Business
7                IT & Software
8         Finance & Accounting
11                    Business
                  ...         
165270                Business
165272     Photography & Video
165281                Business
165286           IT & Software
165288    Teaching & Academics
Name: gaTrackerData.sector, Length: 59565, dtype: object

In [ ]:
glassdoor_data_cleaned["gaTrackerData.sector"].isnull().sum()

0

In [ ]:
glassdoor_data_cleaned.shape

(59565, 163)

## Index

index = 'gaTrackerData.pageRequestGuid.guid' (glassdoor_data)

index = 'id', 'reviews.val.id' (glassdoor_review_data)


In [ ]:
len(glassdoor_data_cleaned['gaTrackerData.jobId.long']) == len(glassdoor_data_cleaned['gaTrackerData.jobId.long'].unique())


False

In [ ]:
glassdoor_data_cleaned_filtered = glassdoor_data_cleaned[glassdoor_useful_cols]
glassdoor_data_cleaned_filtered

gaTrackerData.pageRequestGuid.guid  gaTrackerData.jobId.long  \
0        0000016e56d0b5f2b16c3aea720c55ae              3.227100e+09   
4        0000016e56d0b5ba9a6c4efc3844657c              3.201516e+09   
7        0000016e56d0b8bd9723afa8a8463e86              3.377970e+09   
8        0000016e56d0b8f2906e0190a37a4e65              3.296801e+09   
11       0000016e56d0b91c94a12db296d1fe29              3.328174e+09   
...                                   ...                       ...   
165270   0000016e5781b791aedd42cc0e3921af              3.401740e+09   
165272   0000016e5781b8f5be71e05b5b47fbe1              3.410744e+09   
165281   0000016e5781bb72b09481463ea365b8              3.358602e+09   
165286   0000016e5781bd30b2ccbb057b43bef3              3.301658e+09   
165288   0000016e5781bfcc9ae1ab7c94da1673              3.402587e+09   

                                   gaTrackerData.jobTitle  \
0                      Biogas Project Development Manager   
4                                        Research Manager   
7       Post-Doctoral Research Visit F/M Microscopic d...   
8       Technical Consultant for Microsoft Infrastruct...   
11                                        Program Manager   
...                                                   ...   
165270              Associate Product Manager, Hematology   
165272                                       Data Analyst   
165281                  Application Engineer - Protection   
165286      Data Center Partner Marketing Manager (Japan)   
165288    Opleidingsmanager Industrieel Product Ontwerpen   

       gaTrackerData.location map.country  gaTrackerData.sector  \
0                      Warsaw      Poland  Finance & Accounting   
4                   Bengaluru          IN              Business   
7           Villeneuve-d'Ascq          FR         IT & Software   
8                     Jakarta          ID  Finance & Accounting   
11               Redwood City          US              Business   
...                       ...         ...                   ...   
165270              Vancouver      Canada              Business   
165272          San Francisco          US   Photography & Video   
165281                 Madrid       Spain              Business   
165286                  Tokyo          JP         IT & Software   
165288              Groningen          NL  Teaching & Academics   

        benefits.benefitRatingDecimal  \
0                            0.000000   
4                            0.000000   
7                            0.000000   
8                            0.000000   
11                           0.000000   
...                               ...   
165270                       0.000000   
165272                       3.000000   
165281                       0.000000   
165286                       3.941176   
165288                       0.000000   

                                          job.description  \
0       We are the drivers of the future who offer a g...   
4       <strong>Department:</strong> Research &amp; Op...   
7       Le descriptif de l&rsquo;offre ci-dessous est ...   
8       Technical Consultant will handle solution deli...   
11      <strong>Overview:</strong>\n<br/><br/>\nAs the...   
...                                                   ...   
165270  <h2><strong>Job Description Summary</strong></...   
165272  Job Description<br/><br/>Ubisoft is committed ...   
165281  SEL seeks a experienced Application Engineer -...   
165286  <strong>Job Description</strong><br/><br/>Make...   
165288  Het Instituut voor Engineering is &eacute;&eac...   

                        header.employerName  reviews     job.discoverDate  
0                               Modus Group     2801  2019-10-24T04:52:23  
4               MARKELYTICS SOLUTIONS INDIA     2805  2019-11-01T13:40:48  
7                                     Inria     2808  2019-11-04T12:46:42  
8               Mitra Integrasi Informatika     2809  2019-11-05T04:45:03  
11        

In [ ]:
duplicate_ids = glassdoor_data_cleaned_filtered['gaTrackerData.jobId.long'].value_counts()
duplicate_ids

3.361486e+09    6
3.221354e+09    6
3.386220e+09    5
3.389596e+09    5
3.393104e+09    5
               ..
3.263646e+09    1
3.330680e+09    1
2.858303e+09    1
3.407313e+09    1
3.410744e+09    1
Name: gaTrackerData.jobId.long, Length: 39777, dtype: int64

In [ ]:
len(glassdoor_data_cleaned['reviews']) == len(glassdoor_data_cleaned['reviews'].unique())

True

In [ ]:
duplicate_ids.index[1]

3221353697.0

In [ ]:
glassdoor_data_cleaned_filtered[glassdoor_data_cleaned_filtered['gaTrackerData.jobId.long'] == 3221353697]

gaTrackerData.pageRequestGuid.guid  gaTrackerData.jobId.long  \
32391    0000016e56fbd25f92c639f20decf11e              3.221354e+09   
69149    0000016e5788cab680dcf5d3df4f0a23              3.221354e+09   
74433    0000016e56a35c79b505427c1b40e167              3.221354e+09   
89770    0000016e57e3a13b87d84473eaa3b678              3.221354e+09   
126201   0000016e56078de696e3c91565264e77              3.221354e+09   
138309   0000016e57a8638fbdcd58f33290241d              3.221354e+09   

                                   gaTrackerData.jobTitle  \
32391   Bioinformatics, Scientist / Senior Scientist /...   
69149   Bioinformatics, Scientist / Senior Scientist /...   
74433   Bioinformatics, Scientist / Senior Scientist /...   
89770   Bioinformatics, Scientist / Senior Scientist /...   
126201  Bioinformatics, Scientist / Senior Scientist /...   
138309  Bioinformatics, Scientist / Senior Scientist /...   

       gaTrackerData.location map.country gaTrackerData.sector  \
32391               San Diego          US             Business   
69149               San Diego          US             Business   
74433               San Diego          US             Business   
89770               San Diego          US             Business   
126201              San Diego          US             Business   
138309              San Diego          US             Business   

        benefits.benefitRatingDecimal  \
32391                             0.0   
69149                             0.0   
74433                             0.0   
89770                             0.0   
126201                            0.0   
138309                            0.0   

                                          job.description header.employerName  \
32391   <b>POSITION SUMMARY:</b>\r\n<p>We seek a talen...    Denovo Biopharma   
69149   <b>POSITION SUMMARY:</b>\r\n<p>We seek a talen...    Denovo Biopharma   
74433   <b>POSITION SUMMARY:</b>\r\n<p>We seek a talen...    Denovo Biopharma   
89770   <b>POSITION SUMMARY:</b>\r\n<p>We seek a talen...    Denovo Biopharma   
126201  <b>POSITION SUMMARY:</b>\r\n<p>We seek a talen...    Denovo Biopharma   
138309  <b>POSITION SUMMARY:</b>\r\n<p>We seek a talen...    Denovo Biopharma   

        reviews     job.discoverDate  
32391     16392  2019-10-18T17:58:26  
69149     65060  2019-10-18T17:58:26  
74433    142834  2019-10-18T17:58:26  
89770    130481  2019-10-18T17:58:26  
126201      912  2019-10-18T17:58:26  
138309    88620  2019-10-18T17:58:26

In [ ]:
glassdoor_review_data[glassdoor_review_data['id'] == 16392]

id  index reviews.val.cons reviews.val.date reviews.val.featured  \
181643  16392    NaN              NaN              NaN                  NaN   

        reviews.val.helpfulCount  reviews.val.id reviews.val.pros  \
181643                       NaN             NaN              NaN   

       reviews.val.publishedOn reviews.val.publisher  ...  \
181643                     NaN                   NaN  ...   

        reviews.val.reviewerJobTitle  reviews.val.reviewerLocation  \
181643                           NaN                           NaN   

        reviews.val.reviewerStatus  reviews.val.summaryPoints.ceoApproval  \
181643                         NaN                                    NaN   

        reviews.val.summaryPoints.outlook  \
181643                                NaN   

        reviews.val.summaryPoints.recommend reviews.val.title  \
181643                                  NaN               NaN   

       reviews.val.adviceToManagement reviews.val.companyResponse  \
181643                            NaN                         NaN   

       reviews.val.reviewResponses  
181643                         NaN  

[1 rows x 28 columns]

In [ ]:
glassdoor_review_data_filtered = glassdoor_review_data[glassdoor_review_useful_cols]

In [ ]:
glassdoor_review_data_filtered[glassdoor_review_data_filtered['id'] == 88620]

id  reviews.val.id reviews.val.publishedOn reviews.val.pros  \
336119  88620             NaN                     NaN              NaN   

       reviews.val.cons  
336119              NaN

In [ ]:
glassdoor_review_data_filtered_joined = pd.merge(glassdoor_review_data_filtered, glassdoor_data_cleaned_filtered[['gaTrackerData.jobId.long', 'reviews']], how='left', left_on='id', right_on='reviews').dropna(subset=['gaTrackerData.jobId.long', 'reviews.val.id'])

In [ ]:
# index = 'gaTrackerData.pageRequestGuid.guid'

len(glassdoor_data_cleaned ['gaTrackerData.pageRequestGuid.guid']) ==len(glassdoor_data_cleaned ['gaTrackerData.pageRequestGuid.guid'].unique())
# 59,565 == 59,565

True

In [ ]:
len(glassdoor_review_data_cleaned ['id']) ==len(glassdoor_review_data_cleaned ['id'].unique())


False

In [ ]:
glassdoor_review_data_filtered_joined['reviews.val.id'].value_counts()

30004404.0    282
29906851.0    280
29816017.0    269
29545980.0    269
29546131.0    247
             ... 
453972.0        1
28948382.0      1
17812142.0      1
1754917.0       1
13820804.0      1
Name: reviews.val.id, Length: 41225, dtype: int64

In [ ]:
glassdoor_review_data_filtered_joined[glassdoor_review_data_filtered_joined['reviews.val.id'] == 30004404]

id  reviews.val.id reviews.val.publishedOn  \
158      44066      30004404.0             2 weeks ago   
3505     49407      30004404.0             2 weeks ago   
4479       190      30004404.0             2 weeks ago   
5362     44927      30004404.0             2 weeks ago   
5690     45045      30004404.0             2 weeks ago   
...        ...             ...                     ...   
413149   95307      30004404.0             2 weeks ago   
414798   96353      30004404.0             2 weeks ago   
415206   98113      30004404.0             2 weeks ago   
415891   98371      30004404.0             2 weeks ago   
416402  100561      30004404.0             2 weeks ago   

                    reviews.val.pros                reviews.val.cons  \
158     Great working place to learn  No great opportunities to grow   
3505    Great working place to learn  No great opportunities to grow   
4479    Great working place to learn  No great opportunities to grow   
5362    Great working place to learn  No great opportunities to grow   
5690    Great working place to learn  No great opportunities to grow   
...                              ...                             ...   
413149  Great working place to learn  No great opportunities to grow   
414798  Great working place to learn  No great opportunities to grow   
415206  Great working place to learn  No great opportunities to grow   
415891  Great working place to learn  No great opportunities to grow   
416402  Great working place to learn  No great opportunities to grow   

        gaTrackerData.jobId.long   reviews  
158                 3.403405e+09   44066.0  
3505                3.349932e+09   49407.0  
4479                3.397577e+09     190.0  
5362                3.234591e+09   44927.0  
5690                3.242379e+09   45045.0  
...                          ...       ...  
413149              3.386174e+09   95307.0  
414798              3.319089e+09   96353.0  
415206              3.332126e+09   98113.0  
415891              3.409079e+09   98371.0  
416402              3.400097e+09  100561.0  

[282 rows x 7 columns]

In [ ]:
glassdoor_data_cleaned_filtered = glassdoor_data_cleaned[glassdoor_useful_cols]
glassdoor_data_cleaned_filtered

gaTrackerData.pageRequestGuid.guid  gaTrackerData.jobId.long  \
0        0000016e56d0b5f2b16c3aea720c55ae              3.227100e+09   
4        0000016e56d0b5ba9a6c4efc3844657c              3.201516e+09   
7        0000016e56d0b8bd9723afa8a8463e86              3.377970e+09   
8        0000016e56d0b8f2906e0190a37a4e65              3.296801e+09   
11       0000016e56d0b91c94a12db296d1fe29              3.328174e+09   
...                                   ...                       ...   
165270   0000016e5781b791aedd42cc0e3921af              3.401740e+09   
165272   0000016e5781b8f5be71e05b5b47fbe1              3.410744e+09   
165281   0000016e5781bb72b09481463ea365b8              3.358602e+09   
165286   0000016e5781bd30b2ccbb057b43bef3              3.301658e+09   
165288   0000016e5781bfcc9ae1ab7c94da1673              3.402587e+09   

                                   gaTrackerData.jobTitle  \
0                      Biogas Project Development Manager   
4                                        Research Manager   
7       Post-Doctoral Research Visit F/M Microscopic d...   
8       Technical Consultant for Microsoft Infrastruct...   
11                                        Program Manager   
...                                                   ...   
165270              Associate Product Manager, Hematology   
165272                                       Data Analyst   
165281                  Application Engineer - Protection   
165286      Data Center Partner Marketing Manager (Japan)   
165288    Opleidingsmanager Industrieel Product Ontwerpen   

       gaTrackerData.location map.country  gaTrackerData.sector  \
0                      Warsaw      Poland  Finance & Accounting   
4                   Bengaluru          IN              Business   
7           Villeneuve-d'Ascq          FR         IT & Software   
8                     Jakarta          ID  Finance & Accounting   
11               Redwood City          US              Business   
...                       ...         ...                   ...   
165270              Vancouver      Canada              Business   
165272          San Francisco          US   Photography & Video   
165281                 Madrid       Spain              Business   
165286                  Tokyo          JP         IT & Software   
165288              Groningen          NL  Teaching & Academics   

        benefits.benefitRatingDecimal  \
0                            0.000000   
4                            0.000000   
7                            0.000000   
8                            0.000000   
11                           0.000000   
...                               ...   
165270                       0.000000   
165272                       3.000000   
165281                       0.000000   
165286                       3.941176   
165288                       0.000000   

                                          job.description  \
0       We are the drivers of the future who offer a g...   
4       <strong>Department:</strong> Research &amp; Op...   
7       Le descriptif de l&rsquo;offre ci-dessous est ...   
8       Technical Consultant will handle solution deli...   
11      <strong>Overview:</strong>\n<br/><br/>\nAs the...   
...                                                   ...   
165270  <h2><strong>Job Description Summary</strong></...   
165272  Job Description<br/><br/>Ubisoft is committed ...   
165281  SEL seeks a experienced Application Engineer -...   
165286  <strong>Job Description</strong><br/><br/>Make...   
165288  Het Instituut voor Engineering is &eacute;&eac...   

                        header.employerName  reviews     job.discoverDate  
0                               Modus Group     2801  2019-10-24T04:52:23  
4               MARKELYTICS SOLUTIONS INDIA     2805  2019-11-01T13:40:48  
7                                     Inria     2808  2019-11-04T12:46:42  
8               Mitra Integrasi Informatika     2809  2019-11-05T04:45:03  
11        

In [ ]:
glassdoor_review_data_cleaned_filtered = glassdoor_review_data_cleaned[glassdoor_review_useful_cols]
glassdoor_review_data_cleaned_filtered

id  reviews.val.id reviews.val.publishedOn  \
0        44001        689142.0              2010-10-20   
4        44004      26381884.0              2019-05-29   
5        44004      25136975.0              2019-03-13   
6        44004      24751800.0              2019-02-18   
7        44004      24288425.0              2019-01-19   
...        ...             ...                     ...   
422175  116800      28300687.0              2019-07-29   
422176  116800      24604428.0              2019-02-08   
422177  116800      23444869.0              2018-11-16   
422178  116800      22918057.0              2018-10-12   
422179  116800      22015270.0              2018-08-17   

                                         reviews.val.pros  \
0       Great brand , Good leadership , Clear business...   
4           Very good opportunities to learn technologies   
5        You can learn technically a lot in this company.   
6       - Very friendly environment.\r\n- Highly exper...   
7       Great environment, helpful people, tensionless...   
...                                                   ...   
422175  I read that Capco is too small to compete with...   
422176  Everybody is really helpful. Especially if the...   
422177  As I’m not with Capco for a long time, I’m onl...   
422178  - good working atmosphere\r\n- colleagues will...   
422179  -Quick and efficient onboarding\r\n-Management...   

                                         reviews.val.cons  
0                    Still not big enough in market place  
4       Alot of friends working together which isn't v...  
5       Working hours are not good and need to add the...  
6                                     No Real Cons at all  
7       Little supervision, no precise individuals-fee...  
...                                                   ...  
422175                   None to mention in this context.  
422176     Currently no negatives issues to be mentioned.  
422177  Recruitment:\r\n- Partly uncertainty regarding...  
422178    - few contact with colleagues on other projects  
422179                            None identified so far.  

[259600 rows x 5 columns]

In [ ]:
glassdoor_jobs = glassdoor_data_cleaned_filtered.drop(['gaTrackerData.pageRequestGuid.guid', 'reviews'], axis=1).drop_duplicates()
glassdoor_reviews = glassdoor_review_data_cleaned_filtered.drop(['id'], axis=1).drop_duplicates()
glassdoor_reviews.shape

(66810, 4)

In [ ]:
glassdoor_job_review_relations = pd.merge(glassdoor_data_cleaned_filtered[['gaTrackerData.jobId.long', 'reviews']], glassdoor_review_data_cleaned_filtered[['id', 'reviews.val.id']], how='inner', left_on='reviews', right_on='id')
glassdoor_job_review_relations

gaTrackerData.jobId.long  reviews     id  reviews.val.id
0                   3.201516e+09     2805   2805      22979568.0
1                   3.201516e+09     2805   2805      26430125.0
2                   3.201516e+09     2805   2805      26242080.0
3                   3.201516e+09     2805   2805      24767894.0
4                   3.201516e+09     2805   2805      23810207.0
...                          ...      ...    ...             ...
117886              3.301658e+09    58397  58397      12689371.0
117887              3.402587e+09    58399  58399      18643482.0
117888              3.402587e+09    58399  58399      18391827.0
117889              3.402587e+09    58399  58399      18314631.0
117890              3.402587e+09    58399  58399      22194630.0

[117891 rows x 4 columns]

In [ ]:
glassdoor_reviews[glassdoor_reviews['reviews.val.id'] == 30293890]

Empty DataFrame
Columns: [reviews.val.id, reviews.val.publishedOn, reviews.val.pros, reviews.val.cons]
Index: []

In [ ]:
grouped_review = glassdoor_review_data_cleaned.groupby(["id", "reviews.val.id"])

counts = grouped_review.size()
print(counts)

(counts == 1).all()

id      reviews.val.id
3       29254613.0        1
5       21624925.0        1
        21638140.0        1
6       28158615.0        1
        28289865.0        1
                         ..
165288  12610165.0        1
        23755685.0        1
        24507959.0        1
        25214206.0        1
        25882940.0        1
Length: 259600, dtype: int64


True

## Column rename

In [ ]:
glassdoor_cleaned_final = glassdoor_data_cleaned.rename(columns={
    'gaTrackerData.pageRequestGuid.guid': 'uid',
    'gaTrackerData.jobId.long': 'job_id',
    'gaTrackerData.jobTitle': 'job_title',
    'gaTrackerData.location': 'city',
    'map.country': 'country',
    'gaTrackerData.sector': 'category',
    'benefits.benefitRatingDecimal': 'benefit_rating',
    'job.description': 'description',
    'header.employerName': 'employer_name',
    'reviews': 'num_reviews',
    'job.discoverDate': "date"
    })

In [ ]:
glassdoor_review_cleaned_final = glassdoor_review_data_cleaned.rename(columns= {
    'reviews.val.id': "review_id",
    'reviews.val.publishedOn': 'review_date',
    'reviews.val.pros': 'pros',
    'reviews.val.cons': 'cons'
})

In [ ]:
glassdoor_cleaned_final.shape

(59565, 163)

In [ ]:
glassdoor_review_cleaned_final.shape

(259600, 28)

## Decomposition

**glassdoor_data:**


> Useful attributes:


'uid', 'job_id', 'job_title', 'city', 'country',
'category', 'benefit_rating', 'description', 'employer_name', 'num_reviews', 'date'


> 3NF decomposition:

candidate keys: uid

glassdoor_job ('uid', 'job_id', 'job_title', 'city', 'country', 'category', 'benefit_rating', 'description', 'employer_name', 'num_reviews', 'date')

FD:

uid->job_id

uid->'job_title'

uid->'city'

uid->'country'

uid->'category'

uid->'benefit_rating'

uid->'description'

uid->'employer_name'

uid->'num_reviews'

uid->'date'


**glassdoor_review_data:**


> Useful attributes:


'id', 'review_id', 'review_date', 'pros', 'cons'



> 3NF decomposition:

glassdoor_review ('id', 'review_id', 'review_date', 'pros', 'cons')

candidate keys: "id", "review_id"

FD:

"id", "review_id" -> 'review_date'

"id", "review_id"-> 'pros'

"id", "review_id"-> 'cons'


In [ ]:
table_glassdoor = glassdoor_cleaned_final[['uid', 'job_id', 'job_title', 'city', 'country',
'category', 'benefit_rating', 'description', 'employer_name', 'num_reviews', 'date']].copy()
# table_glassdoor = table_glassdoor.to_csv("/content/drive/My Drive/550_Project_Dataset/glassdoor_job.csv", index = False)

In [ ]:
table_glassdoor_review = glassdoor_review_cleaned_final[['id', 'review_id', 'review_date', 'pros', 'cons']].copy()
# table_glassdoor_review = table_glassdoor_review.to_csv("/content/drive/My Drive/550_Project_Dataset/glassdoor_reviews_cleaned.csv", index= False)